Introduce Finetuning Harness for In-Situ Reinforcement Learning of Agentic Workflows

dnandakumar-nv opened the pull request:
## Summary

This PR introduces a comprehensive **finetuning harness** to the NeMo Agent Toolkit, enabling in-situ reinforcement learning (RL) of agentic LLM workflows. The harness is designed with a decoupled architecture that separates training logic from backends, allowing users to train agents with any RL framework while collecting trajectories workflow evalutations.

### Key Features

- **Three-component architecture**: `TrajectoryBuilder`, `TrainerAdapter`, and `Trainer` interfaces that can be implemented independently
- **Backend-agnostic design**: Train with OpenPipe ART, NeMo Aligner, or custom implementations
- **In-situ training**: Collect trajectories during actual workflow execution, not from offline datasets
- **Curriculum learning**: Progressively introduce harder examples during training
- **GRPO support**: Group trajectories by example for Group Relative Policy Optimization

## What's Included

### 1. Core Finetuning Harness (`src/nat/finetuning/`)

**New Interfaces:**

| Interface | Purpose |
|-----------|---------|
| `TrajectoryBuilder` | Runs evaluations, collects episodes, computes rewards, groups trajectories |
| `TrainerAdapter` | Bridges NAT to external training backends, handles job submission and monitoring |
| `Trainer` | Orchestrates the complete finetuning loop across epochs |

**Data Models (`src/nat/data_models/finetuning.py`):**

- `Trajectory`: Episode sequence with reward and metadata
- `EpisodeItem`: Individual conversation turns with role, content, and logprobs
- `TrajectoryCollection`: Grouped trajectories for GRPO optimization
- `FinetuneConfig`: Complete finetuning configuration
- `CurriculumLearningConfig`: Progressive difficulty settings
- `TrainingJobRef` / `TrainingJobStatus`: Job tracking primitives

**Runtime (`src/nat/finetuning/finetuning_runtime.py`):**

- Async finetuning execution
- Component initialization and binding
- Epoch orchestration with validation support

**CLI Command (`src/nat/cli/commands/finetune.py`):**

```bash
nat finetune --config_file=config.yml \
    --validation_dataset=val.json \
    --validation_interval=5
```

### 2. OpenPipe ART Plugin (`packages/nvidia_nat_openpipe_art/`)

A complete implementation of the finetuning interfaces for [OpenPipe ART](https://art.openpipe.ai/):

| Component | Description |
|-----------|-------------|
| `ARTTrajectoryBuilder` | Parallel evaluation runs, logprob extraction, reward computation |
| `ARTTrainerAdapter` | ART client management, trajectory format conversion, async job submission |
| `ARTTrainer` | Epoch orchestration, curriculum learning, validation, progress visualization |

**Configuration types:**
- `openpipe_art_traj_builder`
- `openpipe_art_trainer_adapter`
- `openpipe_art_trainer`

### 3. Tic-Tac-Toe RL Example (`examples/finetuning/rl_with_openpipe_art/`)

A complete working example demonstrating GRPO training:

- **Task**: LLM plays Tic-Tac-Toe against a random opponent
- **Reward function**: Alpha-beta pruning based position evaluation (continuous, dense rewards)
- **Workflow**: Full game loop with intermediate step tracking
- **Evaluators**: Win percentage and step-value computation
- **Configs**: Pre-training, training, and post-training evaluation

The example includes a sophisticated reward function that:
- Uses game-theoretic minimax with alpha-beta pruning to identify forced wins/losses
- Provides continuous rewards in [0, 1] for non-terminal positions
- Applies temporal discounting to weight earlier moves more heavily
- Adds bonus rewards for winning positions

### 4. Documentation (`docs/source/reference/finetuning/`)

Comprehensive documentation split into focused guides:

| Document | Content |
|----------|---------|
| `index.md` | Overview, architecture diagram, quick start |
| `concepts.md` | RL fundamentals, trajectories, curriculum learning, credit assignment, GRPO/PPO/DPO |
| `extending.md` | Step-by-step guide to implementing custom components |
| `rl_with_openpipe.md` | OpenPipe ART integration details, configuration reference |

### 5. Tests

- Unit tests for finetuning data models
- Integration tests for trajectory building
- End-to-end tests for the ART integration
- Example workflow tests

## Architecture

```
┌─────────────────────────────────────────────────────────────────────────┐
│                              Trainer                                     │
│  (Orchestrates the finetuning loop across epochs)                       │
│                                                                         │
│  ┌───────────────────────┐         ┌───────────────────────────┐       │
│  │  TrajectoryBuilder    │         │    TrainerAdapter         │       │
│  │                       │         │                           │       │
│  │  - Runs evaluations   │ ──────► │  - Validates trajectories │       │
│  │  - Collects episodes  │         │  - Submits to backend     │       │
│  │  - Computes rewards   │         │  - Monitors training      │       │
│  │  - Groups trajectories│         │  - Reports status         │       │
│  └───────────────────────┘         └───────────────────────────┘       │
└─────────────────────────────────────────────────────────────────────────┘
                                         │
                                         ▼
                            ┌─────────────────────────┐
                            │   Remote Training       │
                            │      Backend            │
                            │  (OpenPipe ART, etc.)   │
                            └─────────────────────────┘
```

## Configuration Example

```yaml
llms:
  training_llm:
    _type: openai
    model_name: Qwen/Qwen2.5-3B-Instruct
    base_url: http://localhost:8000/v1

workflow:
  _type: my_workflow
  llm: training_llm

eval:
  general:
    max_concurrency: 16
    dataset:
      _type: json
      file_path: data/training.json
  evaluators:
    reward:
      _type: my_evaluator

trajectory_builders:
  builder:
    _type: openpipe_art_traj_builder
    num_generations: 2

trainer_adapters:
  adapter:
    _type: openpipe_art_trainer_adapter
    backend:
      ip: localhost
      port: 7623
      base_model: Qwen/Qwen2.5-3B-Instruct
    training:
      learning_rate: 1e-6

trainers:
  trainer:
    _type: openpipe_art_trainer

finetuning:
  enabled: true
  trainer: trainer
  trajectory_builder: builder
  trainer_adapter: adapter
  reward_function:
    name: reward
  num_epochs: 20
  curriculum_learning:
    enabled: true
    initial_percentile: 0.3
    expansion_interval: 5
```

## Design Decisions

### Why Decoupled Architecture?

1. **Backend flexibility**: Organizations may have existing training infrastructure (NeMo Customizer, custom GRPO implementations) they want to use
2. **Algorithm agnosticism**: Support GRPO, PPO, DPO, or SFT without code changes
3. **Testability**: Each component can be tested independently with mocks
4. **Future-proofing**: New training backends can be added without modifying core logic

### Why In-Situ Training?

1. **Real data distribution**: Training data reflects actual use cases and edge conditions
2. **Continuous improvement**: Models improve while the system remains operational
3. **Tighter feedback loops**: Hours instead of weeks between training iterations
4. **Adaptive**: Naturally handles distribution shift

## Breaking Changes

None. This is a new feature addition.

## Migration Guide

N/A - New feature.

## Testing

```bash
# Run finetuning harness tests
pytest tests/nat/finetuning/

# Run OpenPipe ART plugin tests
pytest packages/nvidia_nat_openpipe_art/tests/

# Run example tests
pytest examples/finetuning/rl_with_openpipe_art/tests/
```

## Usage

### Running the Example

1. **Start the ART server** (requires GPU):
   ```bash
   source ~/.venvs/art-env/bin/activate
   uv run art --host 0.0.0.0 --port 7623
   ```

2. **Run pre-training baseline**:
   ```bash
   nat eval --config_file=examples/finetuning/rl_with_openpipe_art/src/rl_with_openpipe_art/configs/config_pre_train.yml
   ```

3. **Run finetuning**:
   ```bash
   nat finetune --config_file=examples/finetuning/rl_with_openpipe_art/src/rl_with_openpipe_art/configs/config.yml
   ```

4. **Run post-training evaluation**:
   ```bash
   nat eval --config_file=examples/finetuning/rl_with_openpipe_art/src/rl_with_openpipe_art/configs/config_post_train.yml
   ```

## Checklist

- [x] Core finetuning interfaces implemented
- [x] Finetuning runtime and CLI command
- [x] Data models for trajectories, episodes, and configurations
- [x] OpenPipe ART plugin package
- [x] Tic-Tac-Toe RL example with alpha-beta reward function
- [x] Comprehensive documentation (concepts, extending, ART integration)
- [x] Unit and integration tests
- [x] Example README with step-by-step instructions

## Future Work in other PRs

- Additional training backends (NeMo Customizer)
- DPO/SFT trajectory builder variants
- Checkpoint export utilities


**Reviewers**: Please pay particular attention to:
1. Interface design in `src/nat/finetuning/interfaces/`
2. Data model structure in `src/nat/data_models/finetuning.py`
3. Documentation completeness in `docs/source/reference/finetuning/`


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Experimental finetuning workflow: finetune CLI, OpenPipe ART backend support, new finetuning APIs, and builder/workflow integrations.
  * Example: end-to-end Tic‑Tac‑Toe RL example with configs, evaluator, agents, and runnable finetuning configs.

* **Documentation**
  * New comprehensive finetuning docs (concepts, extending, OpenPipe integration, CLI).

* **Packaging & Examples**
  * Optional OpenPipe ART package and workspace wiring; example added to workspace.

* **Tests**
  * Extensive unit tests for finetuning components, adapters, builders, and parsers.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dnandakumar-nv committed:
Update `uv.lock` to revise dependencies and add overrides

This commit updates the `uv.lock` file to increase the revision number and introduce new package overrides for `litellm` and `openai`. Additionally, metadata like upload times for files has been added, providing a more detailed package resolution record.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Remove OpenPipe ART finetuning backend implementation

Deleted the entire OpenPipe ART finetuning backend, including configurations, runner, and registration modules. This removal simplifies the codebase by eliminating components related to the ART backend that are no longer required.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Remove OpenPipe ART finetuning backend implementation

Deleted the entire OpenPipe ART finetuning backend, including configurations, runner, and registration modules. This removal simplifies the codebase by eliminating components related to the ART backend that are no longer required.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add fine-tuning support with configurable trainer components

This commit introduces a `FinetuningConfig` model and integrates fine-tuning support across the codebase. It includes new components such as `Trainer`, `TrainerAdapter`, and `TrajectoryBuilder`, along with methods to manage and bind these components. Fine-tuning is now fully configurable and enabled through the updated workflow builder and configuration interfaces.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Rename FinetuningRunner to Trainer across the codebase

Replaced all instances of FinetuningRunner with Trainer to reflect updated naming conventions. This change ensures consistency in naming and improves clarity in describing the finetuning workflow orchestration. Updated relevant imports, class definitions, method signatures, and type hints.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add fine-tuning trainer registration and type handling support

Introduced structures and methods to register, retrieve, and handle fine-tuning trainers, trainer adapters, and trajectory builders. Updated type registry and workflows to support these components, ensuring seamless integration into the fine-tuning process.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor training components into distinct groups

Separated `TRAINING` into `TRAINERS`, `TRAINER_ADAPTERS`, and `TRAJECTORY_BUILDERS` for better modularity. Adjusted related models, workflow logic, and configurations to align with the new structure. This improves clarity and supports more granular component handling.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add support for fine-tuning components in workflow builder

Introduced Trainer, TrainerAdapter, and TrajectoryBuilder components with configurations, mocking methods, and dependency tracking. Enhanced testing and registration workflows to validate the integration of these new components into the builder and workflow lifecycle.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add reward config and comprehensive tests for finetuning modules.

This commit introduces a reward field to critical configuration models, including `TrainerConfig` and `TrajectoryBuilderConfig`, enabling enhanced reward-based customization. Additionally, it implements extensive unit tests for `Trainer`, `TrajectoryBuilder`, and `TrainerAdapter`, covering initialization, run workflows, and error handling for robust validation.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add unit tests for NAT finetuning parsers

Introduced comprehensive unit tests for `llama_index_parser`, `langchain_parser`, and `base_parser` under `tests/nat/finetuning/utils/parsers`. These tests cover various edge cases, ensuring proper functionality of content extraction, message parsing, and sequence validation. Added `__init__.py` files to relevant directories for module structure.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Merge remote-tracking branch 'upstream/develop' into finetuning-harness

dnandakumar-nv committed:
Pre-commit checks

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add `finetune` CLI command for workflow finetuning

Introduced a new `finetune` command to the CLI, allowing users to run finetuning on workflows using collected trajectories. This includes options for specifying a configuration file and overriding config values. Additionally, the command integrates with existing CLI infrastructure and logs progress during execution.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
`Refactor finetuning configuration handling`

Introduced a unified `FinetuneConfig` model to replace `FinetuningConfig` and `FinetuneRunConfig`, simplifying finetuning-related data handling. Updated runtime, interfaces, and CLI to leverage the new configuration model, including enhanced validation and improved CLI parameter handling.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add num_epochs to Trainer and enhance finetuning workflow

Introduced a `num_epochs` field to the `FinetuneConfig` model for managing training iterations. Updated `Trainer` and `TrainerAdapter` to simplify initialization, validation, and cleanup processes. Refactored logic for managing configuration and validation datasets, streamlining the finetuning execution.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor imports and clean up redundant code across finetuning modules

Reorganized imports to adhere to consistent style, prioritizing standard and external libraries first, followed by internal ones. Removed unused imports and redundant whitespace for better readability. Simplified code structure and formatting across multiple files to enhance maintainability and clarity.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor imports and clean up redundant code across finetuning modules

Reorganized imports to adhere to consistent style, prioritizing standard and external libraries first, followed by internal ones. Removed unused imports and redundant whitespace for better readability. Simplified code structure and formatting across multiple files to enhance maintainability and clarity.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add OpenPipe ART integration subpackage to NeMo toolkit

This commit introduces a new subpackage, `nvidia-nat-openpipe-art`, for OpenPipe ART integration in the NVIDIA NeMo Agent Toolkit. It includes configuration files such as `pyproject.toml`, licensing information, and placeholder plugin files for future expansion.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add support for nvidia-nat-openpipe-art package

Integrates the nvidia-nat-openpipe-art package into the project, including its dependency on openpipe-art version 0.5.1. Updates uv.lock and pyproject.toml to reflect the new package and its associated configurations.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add ART trajectory builder implementation for OpenPipe backend

This commit introduces the ARTTrajectoryBuilder and its configuration, enabling trajectory generation for the OpenPipe backend. It includes registration logic, main functionalities, and associated dependency updates.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add ARTTrainerAdapter and enhance OpenPipe ART framework

Introduced ARTTrainerAdapter for integrating with the ART backend, enabling advanced training workflows. Updated configuration classes, registration logic, and initialization to support new functionality, including reward functions in both trajectory builder and trainer adapter.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add ARTTrainer implementation and curriculum learning support

Introduced ARTTrainer for OpenPipe ART backend, enabling finetuning with curriculum learning and trajectory-based rewards. Updated configuration, registration, and dependencies to support these functionalities, including matplotlib for visualizations.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update default trainer run name in configuration

Changed the default value of the `name` field in the configuration from "trainer_run_4" to "trainer_run". This simplifies the default naming convention for trainer runs.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add tests for ARTTrainer and ARTTrainerAdapter functionality

Introduced comprehensive unit tests for `ARTTrainer` and `ARTTrainerAdapter` to validate initialization, training, trajectory handling, status management, and error handling. These tests ensure robust functionality and resilience of training components.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add RL with OpenPipe ART Tic-Tac-Toe workflow and components

Introduces a Tic-Tac-Toe game workflow utilizing LangChain for LLM-based agents. Includes game logic, XML move parsing, heuristic evaluation, and a configurable function for LLM interaction. Adds dependency management, configuration, and registration setup for OpenPipe ART integration.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Revert uv.lock file to revision 1 and clean up metadata.

Rolled back `uv.lock` file from revision 3 to revision 1. Removed redundant `upload-time` metadata for sdist and wheel dependencies to simplify the file structure and ensure consistency.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add missing import for workflow function registration

This ensures that the `rl_with_openpipe_art_function` is properly imported, triggering its registration as expected. This change improves the reliability of the registration process.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add target_model filter to trajectory processing

Introduced a `target_model` attribute to allow filtering of intermediate steps by model name during trajectory processing. Updated the trajectory builder logic to skip steps not matching the specified target model, ensuring more precise fine-tuning when multiple models are involved.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add ruff ignore flags and placeholder data file

Added `noqa` flags to suppress unused import warnings for ruff and flake8. Introduced a new placeholder data file, `data.json`, with sample question-answer pairs for testing purposes.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add custom accuracy evaluator for RL workflow outputs

Introduce `AccuracyEvaluator` class to evaluate RL outputs based on specific scoring logic. Register the evaluator for use in workflows and include its configuration. Also, update logging in the Tic-Tac-Toe workflow from `info` to `debug` for reduced verbosity.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add randomized move option and logging improvements

Introduced a `choose_random` flag for LLM-based Tic Tac Toe players, enabling randomized moves when specified. Enhanced logging by replacing print statements with logger calls and adjusted configuration defaults for better evaluation control.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add finetuning configuration and update annotation handling

Introduced a comprehensive finetuning configuration to enable specific training setups, including trajectory builders, trainer adapters, and reward functions. Updated `config.py` to include new annotations and rebuild handling for trainers, trainer adapters, and trajectory builders. These changes facilitate flexible and scalable finetuning workflows.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add support for new trainer-related keys and handle TypeError

Extend the config validator to support trainers, trainer adapters, and trajectory builders in the GlobalTypeRegistry. Additionally, update the component discovery utility to handle TypeError for types that don’t support isinstance checks, ensuring robust handling of complex type unions.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Enable fine-tuning and update required configuration fields.

Fine-tuning is now activated by default in the configuration file. Removed "config_file" as a required field for validation and updated trainer_adapter references for consistency. These changes streamline the setup process for fine-tuning workflows.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor payload handling and clean up unused code.

Introduce `payload` in `IntermediateStep` for raw generation data. Remove unused methods, redundant error handling, and unused imports, simplifying callback handler logic for better maintainability.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update temperature settings, add component registrations, and fix tests

Adjusted temperature values in the RL configuration for tuning. Introduced registration for finetuning components and added relevant imports in test files. Enhanced the `dask_cluster_fixture` for thread-based workers, improving test reliability and performance.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor: Remove heuristic evaluation logic and update configs

Removed the `evaluate_board_for_player` function and related references to streamline the core logic of the RL module. Updated model configurations in `config.yml` to use new model versions for better alignment with the latest LLM capabilities. These changes simplify the codebase and enhance maintainability.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor RL configs into modular files for clarity.

Split the monolithic RL fine-tuning config into three modular files: baseline, training, and post-training. This improves organization, readability, and distinction between different phases of the workflow.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add heuristic board evaluation and cleanup config files

Introduced an Alpha-Beta-style heuristic evaluation for Tic Tac Toe boards, improving AI decision-making with detailed scoring metrics. Removed outdated configuration YAML files related to fine-tuning and replaced them with cleaner structures. This refactor improves code clarity, reduces redundancy, and enhances AI gameplay logic.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add configs and refine RL workflow for OpenPipe ART

Introduce pre-train, finetune, and post-train configuration files for RL with OpenPipe ART, enabling LLM-specific workflows and evaluations. Refine code by removing heuristic evaluations and streamlining move evaluation logic for improved clarity and maintainability.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Simplify config structures and update default parameters.

Replaced default_factory with None for clarity in backend configs. Adjusted file paths and updated temperature values, model names, and various training parameters for consistency and better evaluation alignment. Removed redundant settings to streamline configuration files.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor player logic and evaluation in RL workflow.

Updated the RL workflow to handle player model configurations more flexibly, simplified move retries, and refined game termination handling. Added failure penalty in evaluation, adjusted default learning rates, and updated dataset answers to reflect new scoring outputs.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Simplify message tracking and improve evaluation logic

Modified parsing logic to prevent duplicate messages, improved evaluation logic for scoring aborted games, and adjusted configurations for training, evaluation, and model setup. Additional enhancements include refining message history handling and making the board display more user-friendly.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update trainer logic and configuration defaults

Adjusted "finish_reason" in trainer adapter to "stop" for consistency and updated training beta default to 0 for improved behavior. Additionally, set max_parser_retries default to 0 in RL example to align with updated retry strategy.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor configurations and logic for improved evaluation.

Updated configuration files to adjust temperature, concurrency, and dataset paths. Refined game logic to account for player roles. Simplified trajectory message handling and expanded evaluation dataset with additional examples.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor RL evaluator and game logic for improved traceability

Updated the RL evaluator to incorporate intermediate step value computation for richer insights during evaluation. Refactored game logic to enable step-level state tracking with value assignments, improving training feedback. Adjusted configurations for model, evaluator, and workflow parameters to align with the updated logic.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Adjust training configs and expand dataset for RL module

Updated training parameters, including learning rate, temperature, and number of epochs. Adjusted naming conventions for clarity and added `max_parser_retries` to configs. Expanded the dataset with additional entries to improve model training coverage.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add finetuning documentation and OpenPipe ART integration

This update introduces documentation on the finetuning harness, covering key concepts, reinforcement learning approaches, and curriculum learning. Additionally, it includes a reference for integrating OpenPipe ART for training LLMs using GRPO optimization. Configuration and CLI usage details are also provided.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add RL Tic-Tac-Toe example using OpenPipe ART

This commit introduces a detailed example demonstrating reinforcement learning finetuning with OpenPipe ART to enhance an LLM's Tic-Tac-Toe performance. Includes a comprehensive README explaining setup, workflow, reward function, and best practices, alongside necessary configs, scripts, and assets.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Merge remote-tracking branch 'upstream/develop' into finetuning-harness

dnandakumar-nv committed:
Update comment for dependency overrides in pyproject.toml

Clarified the purpose of overriding dependencies to prevent regressions caused by nvidia-nat-openpipe-art. Added additional context to emphasize maintaining the status quo for other packages.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Fix relative link paths to Custom Evaluators in docs

Updated the relative paths to the Custom Evaluators page to ensure proper navigation in the documentation. This resolves incorrect linking issues in the finetuning reference guides.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Fix typo in 'logprob' vocabulary entry in Vale config

Corrected the case-insensitivity pattern for 'logprob' to ensure proper matching in the Vale style vocabulary. This update aligns with consistent handling of vocabulary terms.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor module naming and improve setup instructions

Updated module and project naming to align with the "nat" prefix convention. Adjusted logging suppression prefix and enhanced setup instructions in README for better clarity, including specifying virtual environment and dependencies.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor module naming and improve setup instructions

Updated module and project naming to align with the "nat" prefix convention. Adjusted logging suppression prefix and enhanced setup instructions in README for better clarity, including specifying virtual environment and dependencies.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update README for ART setup with corrected Python version and package

Changed the Python version to 3.11 and specified the exact OpenPipe ART package version for consistent setup. This ensures compatibility and avoids potential installation issues.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

willkill07 commented on packages/nvidia_nat_openpipe_art/LICENSE-3rd-party.txt:
Should be a symlink to parent package

willkill07 commented on packages/nvidia_nat_openpipe_art/LICENSE.md:
Should be a symlink to parent package

dnandakumar-nv committed:
Enhance token usage tracking and tool output handling.

Refactor token usage extraction to capture additional details like cached and reasoning tokens. Add robust handling of tool outputs in callbacks, improving support for server-side tool usage and error resilience during schema parsing.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add finetune command documentation to CLI reference

This update introduces the `nat finetune` command, providing detailed documentation for its usage, options, and examples. The finetune command allows in-situ reinforcement learning for iterative agent improvement, incorporating experimental features and extensive configuration options.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dagardner-nv commented on packages/nvidia_nat_openpipe_art/pyproject.toml:
Don't forget to add this new package as a dependency in `packages/nvidia_nat_all/pyproject.toml`

dnandakumar-nv committed:
Merge remote-tracking branch 'upstream/develop' into finetuning-harness

# Conflicts:
#	uv.lock

dnandakumar-nv committed:
Add support for `openpipe-art` package and related dependencies

Integrated `openpipe-art` package and its dependencies, including `litellm` and `openai`. Updated project metadata, extras, and lockfile entries to enable its functionality. This enhancement introduces new examples and configurations for fine-tuning workflows.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update OpenPipe ART integration and dependencies

Updated the description for the OpenPipe ART subpackage in `pyproject.toml` to reflect proper integration context. Added `nvidia-nat-openpipe-art` as a dependency in the `nvidia_nat_all` package for consistency in workspace management.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add guideline to update dependencies in `nvidia-nat-all`

Clarifies that when a new package is added, its name must be included as a dependency in the `nvidia-nat-all` package's `pyproject.toml` file. Ensures consistency in dependency management.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add ADK framework support and shared parser utilities

Introduced ADK-specific parser handling logic for LLM messages and integrated it into the existing framework switch. Refactored common parsing utilities for reusability, reducing code duplication in LangChain and LlamaIndex parsers. Added comprehensive unit tests for ADK parser functionality, ensuring proper content extraction and message parsing across edge cases.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refine formatting, comments, and docstrings for consistency.

Updated bullet point styles in comments for uniformity across files. Standardized docstring formatting, ensuring proper punctuation and clear argument descriptions. These changes improve readability and maintain documentation consistency.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Merge remote-tracking branch 'upstream/develop' into finetuning-harness

willkill07 reviewed (approved):
Minor nits that need to be corrected before merge, but LGTM overall.

dnandakumar-nv committed:
Add licensing files for NVIDIA package dependencies

This commit introduces the Apache License 2.0 and third-party license files to the NVIDIA package. These files provide compliance with open-source licensing requirements by specifying the terms and conditions of use, reproduction, and distribution for the package and its dependencies.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv commented:
/ok to test c6d8565efe89b195b9462a1ec041ba372c0440a7

dnandakumar-nv committed:
Remove outdated test README and add missing SPDX headers

Deleted the `tests/README.md` file as it was outdated and redundant. Added SPDX license headers to multiple config files and updated file paths in configuration files for improved clarity. Also included new terms in the vocabulary acceptance list.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add new path patterns for openpipe_art to path checks

Updated the path checks script to include rules for

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv commented:
/ok to test 1e69989d4a52b797bb5c92c42e08d25351046116

dnandakumar-nv committed:
Fix typo in vocabulary list for "finetuning" and "finetunable"

Corrected capitalization issues in the listed terms to align with consistent styling. This ensures better accuracy and functionality in text validation tools.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv commented:
/ok to test b765ef066ff668218fbad336ed48c22335caac1b

dnandakumar-nv committed:
Refine path checks and update vocabulary/guide.

Improved regex patterns in path checks for better accuracy and coverage. Expanded accepted vocabulary in Vale configurations with finetuning-related terms. Added path-check skip markers in the finetuning documentation for clarity and alignment with new rules.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv commented:
/ok to test a0dc798eea37041ee44ca3d5dd2885834716168f

dnandakumar-nv committed:
Update vocab and refine headers in RL fine-tuning docs

Adjusted vocabulary to include "subsampling" and fix case typo for "rollout". Renamed headers for clarity in documentation, providing consistent class references for `ARTTrainerAdapter` and `ARTTrainer`.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Expand path checks with additional config patterns

Updated the path checks to include a specific config.yaml file path under the `src/rl_with_openpipe_art/configs` directory. This ensures proper handling and validation of the newly added configuration structure.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv commented:
/ok to test a7940d6082f2274aa733e1bf2537f553c436e195

dnandakumar-nv committed:
Add 'Qwen/Qwen.*' pattern to path checks

This update includes the 'Qwen/Qwen.*' regex pattern in the path checks to expand the validation ruleset. It ensures proper handling of paths related to the newly added category.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv commented:
/ok to test a293ed9ee1abb024a54785e5fb68ab84becc701e

dnandakumar-nv commented:
/merge
